File: test/desktop_mode_shim_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:webspace/services/desktop_mode_shim.dart';
import 'package:webspace/services/user_agent_classifier.dart';

void main() {
  group('buildDesktopModeShim', () {
    test('emits the reentrance guard so repeat frames do not re-wrap', () {
      // WebKit / Android WebView re-run initialUserScripts on every frame
      // load. Without this guard the matchMedia wrapper would wrap itself
      // and recurse infinitely.
      final js = buildDesktopModeShim(firefoxLinuxDesktopUserAgent);
      expect(js, contains('__ws_desktop_shim__'));
    });

    test('patches the core navigator/window surfaces', () {
      // If any of these drop out silently, the toggle stops working on
      // sites that gate on just one signal.
      final js = buildDesktopModeShim(firefoxWindowsDesktopUserAgent);
      expect(js, contains('userAgentData'));
      expect(js, contains('maxTouchPoints'));
      expect(js, contains('ontouchstart'));
      expect(js, contains('matchMedia'));
      expect(js, contains('viewport'));
      expect(js, contains('MutationObserver'));
    });

    test('makes navigator.userAgentData undefined (Firefox-style)', () {
      // Our spoofed UA is Firefox-shaped, and Firefox does not implement
      // the Client Hints API. Sites feature-detecting `userAgentData`
      // should see undefined — anything else is a tell.
      final js = buildDesktopModeShim(firefoxMacosDesktopUserAgent);
      expect(js, contains(
        "def('userAgentData', asNative(function() { return undefined; }, 'userAgentData'));"));
    });

    test('Linux UA emits "Linux x86_64" navigator.platform', () {
      final js = buildDesktopModeShim(firefoxLinuxDesktopUserAgent);
      expect(js, contains('"Linux x86_64"'));
    });

    test('macOS UA emits "MacIntel" navigator.platform', () {
      final js = buildDesktopModeShim(firefoxMacosDesktopUserAgent);
      expect(js, contains('"MacIntel"'));
    });

    test('Windows UA emits "Win32" navigator.platform', () {
      final js = buildDesktopModeShim(firefoxWindowsDesktopUserAgent);
      expect(js, contains('"Win32"'));
    });

    test('different UAs emit different shim payloads', () {
      // A refactor that parameterizes one platform value but forgets
      // another would collapse two shims to the same source — guard
      // against that.
      final linux = buildDesktopModeShim(firefoxLinuxDesktopUserAgent);
      final macos = buildDesktopModeShim(firefoxMacosDesktopUserAgent);
      final windows = buildDesktopModeShim(firefoxWindowsDesktopUserAgent);
      expect(linux, isNot(equals(macos)));
      expect(linux, isNot(equals(windows)));
      expect(macos, isNot(equals(windows)));
    });

    test('matchMedia wrapper covers pointer/hover desktop flips', () {
      final js = buildDesktopModeShim(firefoxLinuxDesktopUserAgent);
      // The wrapper must synthesize matches for the desktop side of each
      // pointer/hover query and non-matches for the mobile side; otherwise
      // responsive CSS still picks the mobile layout.
      expect(js, contains('pointer'));
      expect(js, contains('hover'));
      expect(js, contains('fine'));
      expect(js, contains('coarse'));
    });

    test('viewport rewrite uses a desktop-width content value', () {
      final js = buildDesktopModeShim(firefoxLinuxDesktopUserAgent);
      expect(js, contains('width=1280'));
    });

    test('does NOT spoof devicePixelRatio', () {
      // DPR is orthogonal to desktop-vs-mobile layout. Real retina
      // MacBooks and 4K Chrome desktops report dpr >= 2; pinning to 1
      // would downgrade image quality on any modern phone/tablet for no
      // layout benefit. Regression guard against re-introducing the
      // patch — match the actual `Object.defineProperty` shape, not the
      // word `devicePixelRatio` (which appears in the explanatory
      // comment).
      final js = buildDesktopModeShim(firefoxLinuxDesktopUserAgent);
      expect(js, isNot(contains("def(window, 'devicePixelRatio'")));
      expect(js, isNot(contains("'devicePixelRatio',")));
    });
  });
}

File: test/js/desktop_mode_shim.test.js
// Behavioural tests for the desktop-mode shim
// (lib/services/desktop_mode_shim.dart).
//
// Goal: prove the shim actually mutates the JS surface a fingerprinter
// would inspect, not just that the source contains the right substrings —
// the existing test/desktop_mode_shim_test.dart already covers string
// matching. These tests run the dumped fixture inside jsdom and assert
// the post-injection navigator/window state.

const test = require('node:test');
const assert = require('node:assert/strict');
const { loadShim, makeDom, runInDom, readFixture } = require('./helpers/load_shim');

test('Linux fixture sets navigator.platform to "Linux x86_64"', () => {
  const dom = loadShim('desktop_mode/linux.js');
  assert.equal(dom.window.navigator.platform, 'Linux x86_64');
});

test('macOS fixture sets navigator.platform to "MacIntel"', () => {
  const dom = loadShim('desktop_mode/macos.js');
  assert.equal(dom.window.navigator.platform, 'MacIntel');
});

test('Windows fixture sets navigator.platform to "Win32"', () => {
  const dom = loadShim('desktop_mode/windows.js');
  assert.equal(dom.window.navigator.platform, 'Win32');
});

test('navigator.userAgentData is undefined (Firefox-shaped UA)', () => {
  const dom = loadShim('desktop_mode/linux.js');
  // Firefox does not implement Client Hints; sites feature-detecting
  // userAgentData must see undefined, not a Chromium-WebView-populated
  // mobile object.
  assert.equal(dom.window.navigator.userAgentData, undefined);
});

test('navigator.maxTouchPoints is forced to 0', () => {
  const dom = loadShim('desktop_mode/linux.js');
  assert.equal(dom.window.navigator.maxTouchPoints, 0);
});

test('window.ontouchstart is undefined after the shim runs', () => {
  const dom = loadShim('desktop_mode/linux.js');
  // The "in" check is what touch detection libraries use.
  assert.equal(dom.window.ontouchstart, undefined);
});

test('matchMedia(pointer: fine) → matches=true; (pointer: coarse) → false', () => {
  const dom = loadShim('desktop_mode/linux.js');
  const fine = dom.window.matchMedia('(pointer: fine)');
  const coarse = dom.window.matchMedia('(pointer: coarse)');
  assert.equal(fine.matches, true, 'pointer: fine should match on desktop');
  assert.equal(coarse.matches, false, 'pointer: coarse should NOT match');
  assert.equal(fine.media, '(pointer: fine)');
});

test('matchMedia(hover: hover) → matches=true; (hover: none) → false', () => {
  const dom = loadShim('desktop_mode/linux.js');
  assert.equal(dom.window.matchMedia('(hover: hover)').matches, true);
  assert.equal(dom.window.matchMedia('(hover: none)').matches, false);
});

test('matchMedia falls through for non-pointer/hover queries', () => {
  // Width-based queries must not be hijacked — the shim must only forge
  // pointer/hover responses, not affect responsive breakpoints.
  const dom = loadShim('desktop_mode/linux.js');
  const result = dom.window.matchMedia('(min-width: 100px)');
  // jsdom's stub returns { matches: false, media }. The shim should leave
  // it alone (and the wrapper doesn't synthesise a response for this
  // query), so we should get the underlying jsdom result back.
  assert.equal(typeof result.matches, 'boolean');
  assert.equal(result.media, '(min-width: 100px)');
});

test('synthetic matchMedia result has addEventListener / removeEventListener', () => {
  // CSS-in-JS libraries (emotion, styled-components media-query helpers)
  // call addEventListener('change', ...) on MediaQueryList. If the
  // synthetic wrapper drops these methods the page throws.
  const dom = loadShim('desktop_mode/linux.js');
  const fine = dom.window.matchMedia('(pointer: fine)');
  assert.equal(typeof fine.addEventListener, 'function');
  assert.equal(typeof fine.removeEventListener, 'function');
  assert.equal(typeof fine.addListener, 'function');
  assert.equal(typeof fine.removeListener, 'function');
  // Should not throw when a listener is attached/detached.
  const listener = () => {};
  fine.addEventListener('change', listener);
  fine.removeEventListener('change', listener);
});

test('existing <meta name="viewport"> is rewritten to width=1366', () => {
  // A site shipping `width=device-width` defeats useWideViewPort. The
  // shim rewrites every viewport meta — including ones present at
  // injection time — to width=1366, initial-scale=1.0. 1366 clears
  // Bluesky's `(min-width: 1300px)` desktop breakpoint; a smaller
  // value would ship the tablet layout on Android.
  const html = `<!doctype html><html><head>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  </head><body></body></html>`;
  const dom = makeDom({ html });
  runInDom(dom, readFixture('desktop_mode/linux.js'));
  const meta = dom.window.document.querySelector('meta[name="viewport"]');
  assert.ok(meta);
  assert.equal(meta.getAttribute('content'), 'width=1366, initial-scale=1.0');
});

test('viewport meta added later is rewritten via MutationObserver', async () => {
  const dom = loadShim('desktop_mode/linux.js');
  const meta = dom.window.document.createElement('meta');
  meta.setAttribute('name', 'viewport');
  meta.setAttribute('content', 'width=device-width');
  dom.window.document.head.appendChild(meta);
  // MutationObserver callbacks are queued as microtasks. Yield to let
  // them flush.
  await new Promise((resolve) => setImmediate(resolve));
  assert.equal(meta.getAttribute('content'), 'width=1366, initial-scale=1.0');
});

test('re-entrance guard: running the shim twice is a no-op', () => {
  // WebKit and Android WebView re-run initialUserScripts on every frame
  // load. Without the __ws_desktop_shim__ guard, the matchMedia wrapper
  // would wrap itself and recurse forever the second time around.
  const dom = loadShim('desktop_mode/linux.js');
  runInDom(dom, readFixture('desktop_mode/linux.js'));
  // matchMedia should still produce the right answer (not blow the stack
  // or return an unexpected shape).
  assert.equal(dom.window.matchMedia('(pointer: fine)').matches, true);
  assert.equal(dom.window.__ws_desktop_shim__, true);
});
